src/mscorlib/shared/System/Convert.cs

```diff
@@ -2658,83 +2658,132 @@ public static bool TryFromBase64String(string s, Span<byte> bytes, out int bytes
+                        }
+
+                        tempBuffer = tempBuffer.Slice(0, charsWritten);
+                        if (!Base64.TryDecodeFromUtf16(tempBuffer, bytes, out int consumedFromTempBuffer, out int bytesWrittenFromTempBuffer))
+                        {
+                            bytesWritten = default;
+                            return false;
+                        }
+                        bytesWritten += bytesWrittenFromTempBuffer;
+                        chars = chars.Slice(consumedFromChars);
+                        bytes = bytes.Slice(bytesWrittenFromTempBuffer);
+
+                        if ((bytesWrittenFromTempBuffer % 3) != 0)
+                        {
+                            // If we got here, this decode contained one or more padding characters ('='). We can accept trailing whitespace after this
+                            // but nothing else.
+                            for (int i = 0; i < chars.Length; i++)
+                            {
+                                if (!chars[i].IsSpace())
+                                {
+                                    bytesWritten = default;
+                                    return false;
+                                }
+                            }
+                            return true;
+                        }
+
+                        // We now loop again to decode the next run of non-space characters. 
+                    }
                 }
-                return true;
             }
+
+            return true;
         }
 
-        private static unsafe bool TryFromBase64CharsSlow(ReadOnlySpan<char> chars, Span<byte> bytes, out int bytesWritten)
+        [StructLayout(LayoutKind.Sequential)]
+        private struct TempCharBuffer
         {
-            if (chars.Length == 0)
-            {
-                bytesWritten = 0;
-                return true;
-            }
+            public char C0;
+            public char C1;
+            public char C2;
+            public char C3;
+        }
 
-            // We need to get rid of any trailing white spaces.
-            // Otherwise we would be rejecting input such as "abc= ":
-            while (chars.Length > 0)
-            {
-                char lastChar = chars[chars.Length - 1];
-                if (lastChar != ' ' && lastChar != '\n' && lastChar != '\r' && lastChar != '\t')
-                {
-                    break;
-                }
-                chars = chars.Slice(0, chars.Length - 1);
-            }
+        private static void CopyToTempBufferWithoutWhiteSpace(ReadOnlySpan<char> chars, Span<char> tempBuffer, out int consumed, out int charsWritten)
+        {
+            Debug.Assert(tempBuffer.Length != 0); // We only bound-check after writing a character to the tempBuffer.
 
-            fixed (char* charsPtr = &MemoryMarshal.GetReference(chars))
+            charsWritten = 0;
+            for (int i = 0; i < chars.Length; i++)
```

Comment:
Should this start from 1, since we already know that chars[0] is not a whitespace character?

Reply:
Sharing an assumption across a helper method boundary is fragile - I don't think that's worth it. This is already the "oops - this is the slowest path since we don't expect to encounter it often in real usage" path.